File: Week2/Stack_LinkedList.cpp
/*
 * Implement stack data structure with linked list.
 * Works only for data types with = operation defined.
 * You may also overload = operator in order to use stack defined here.
 */
#include <iostream>
using namespace std;

template < typename valType >
class StackIterator;

template < typename valType >
class Node
{
    public:
        valType val;
        Node<valType> *next;
};

template < typename valType >
class Stack
{
    friend class StackIterator<valType>;
    public:
        Stack();
        void push( valType );
        valType pop();
        bool isEmpty() const { return _count == 0; };
        int size() const { return _count; };

        typedef StackIterator<valType> iterator;

        iterator begin() { return iterator( _head ); };
        iterator end() { return iterator( _tail ); };

    private:
        Node<valType> *_head, *_tail;;
        unsigned _count;
};

template < typename valType >
Stack<valType>::Stack()
{
    _head = _tail = NULL;
    _count = 0;
}

template < typename valType >
void Stack<valType>::push( valType elem )
{
    Node<valType> *ptr = new Node<valType>;
    ptr->val = elem;
    ptr->next = _head;
    _head = ptr;
    _count++;
}

template < typename valType >
valType Stack<valType>::pop( )
{
    Node<valType> *ptr = _head;
    _head = _head->next;
    valType tmp = ptr->val;
    delete ptr;
    return tmp;
}

template < typename valType >
class StackIterator
{
    public:
        StackIterator( Node<valType> *ptr ) : _nodePtr(ptr) {} ;

        valType operator*() const { return _nodePtr->val; };
        StackIterator<valType> operator++(int)
        {
            _nodePtr = _nodePtr->next;
            return *this;
        };
        bool operator==( StackIterator<valType> &rhs )
        { return _nodePtr == rhs._nodePtr; };
        bool operator!=( StackIterator<valType> &rhs )
        { return !(*this==rhs); }

    private:
        Node<valType> *_nodePtr;

};


int main()
{
    Stack<int> s;
    s.push(1);
    s.push(2);

    Stack<int>::iterator it1 = s.begin();
    Stack<int>::iterator it2 = s.end();
    Stack<int>::iterator it3 = s.end();
    for ( ; it1 != it2; it1++ )
        cout << *it1 << endl;
}

File: Week2/Stack_LinkedList.java
import java.util.Iterator;
import java.util.NoSuchElementException;

public class Stack_LinkedList<Item> implements Iterable<Item>
{
    private class Node<Item>
    {
        private Item val;
        private Node next;
        private Node ( ) { next = null; };
        private Node ( Item item ) { val = item; next = null; };
    }

    private class ListIterator<Item> implements Iterator<Item>
    {
        private Node<Item> current;
        public ListIterator( Node<Item> node ) { current = node; };

        public boolean hasNext( ) { return current != null; };

        public Item next()
        {
            if ( !hasNext() ) throw new NoSuchElementException();
            Node<Item> n = current;
            current = current.next;
            return n.val;
        }
    }

    Node<Item> head;
    int count;
    public Stack_LinkedList ()
    {
        head = null;
        count = 0;
    }

    public void push( Item elem )
    {
        Node<Item> newhead = new Node<Item>( elem );
        newhead.next = head;
        head = newhead;
        count++;
    }

    public Item pop()
    {
        if ( isEmpty() )
            throw new NoSuchElementException("Stack_LinkedList underflow");
        Node<Item> n = head;
        head = head.next;
        count--;
        return n.val;
    }

    public boolean isEmpty()
    {
        return head == null;
    }

    public Iterator<Item> iterator()
    {
        return new ListIterator<Item>( head );
    }

    public static void main(String[] args)
    {
        Stack_LinkedList<Integer> s = new Stack_LinkedList<>();
        s.push(1);
        s.push(2);
        s.push(4);
        s.push(6);

        for ( int i : s )
            System.out.println(i);
    }
}
